Mathlib/GroupTheory/Perm/Cycle/Basic.lean

```diff
@@ -985,3 +987,154 @@ theorem sum_mul_sum_eq_sum_perm (hσ : σ.IsCycleOn s) (f g : ι → α) :
+    Commute g c ↔
+      ∃ hc' : ∀ x : α, x ∈ c.support ↔ g x ∈ c.support,
+        subtypePerm g hc' ∈ Subgroup.zpowers c.subtypePermOfSupport := by
+  constructor
+  · intro hgc
+    let hgc' := mem_support_iff_of_commute hgc
+    use hgc'
+    obtain ⟨a, ha⟩ := IsCycle.nonempty_support hc
+    obtain ⟨i, hi⟩ := hc.sameCycle (mem_support.mp ha) (mem_support.mp ((hgc' a).mp ha))
+    use i
+    ext ⟨x, hx⟩
+    simp only [subtypePermOfSupport, Subtype.coe_mk, subtypePerm_apply]
+    rw [subtypePerm_apply_zpow_of_mem]
+    obtain ⟨j, rfl⟩ := hc.sameCycle (mem_support.mp ha) (mem_support.mp hx)
+    simp only [← mul_apply, Commute.eq (Commute.zpow_right hgc j)]
+    rw [← zpow_add, add_comm i j, zpow_add]
+    simp only [mul_apply, EmbeddingLike.apply_eq_iff_eq]
+    exact hi
+  · -- converse
+    rintro ⟨hc', h⟩
+    obtain ⟨i, hi⟩ := h
+    ext x
+    simp only [coe_mul, Function.comp_apply]
+    by_cases hx : x ∈ c.support
+    · -- hx : x ∈ c.support
+      suffices hi' : ∀ x ∈ c.support, g x = (c ^ i) x by
+        rw [hi' x hx, hi' (c x) (apply_mem_support.mpr hx)]
+        simp only [← mul_apply, ← zpow_add_one, ← zpow_one_add, add_comm]
+      -- proof of hi'
+      intro x hx
+      let hix := Perm.congr_fun hi ⟨x, hx⟩
+      simp only [← Subtype.coe_inj, subtypePermOfSupport, Subtype.coe_mk, subtypePerm_apply,
+        subtypePerm_apply_zpow_of_mem] at hix
+      exact hix.symm
+    · -- hx : x ∉ c.support
+      rw [not_mem_support.mp hx, eq_comm, ← not_mem_support]
+      contrapose! hx
+      exact (hc' x).mpr hx
+
+/-- A permutation `g` commutes with a cycle `c` if and only if
+  `c.support` is invariant under `g`, and `g` acts on it as a power of `c`. -/
+theorem IsCycle.commute_iff {g c : Perm α} (hc : c.IsCycle) :
+    Commute g c ↔
+      ∃ hc' : ∀ x : α, x ∈ c.support ↔ g x ∈ c.support,
+        ofSubtype (subtypePerm g hc') ∈ Subgroup.zpowers c := by
+  rw [IsCycle.commute_iff' hc]
+  apply exists_congr
+  intro hc'
+  simp only [Subgroup.mem_zpowers_iff]
+  apply exists_congr
+  intro k
+  unfold subtypePermOfSupport
+  rw [subtypePerm_zpow c k]
+  simp only [Equiv.ext_iff, subtypePerm_apply, Subtype.mk.injEq, Subtype.forall]
+  apply forall_congr'
+  intro a
+  by_cases ha : a ∈ c.support
+  · rw [imp_iff_right ha]
+    apply Eq.congr rfl
+    apply symm
+    exact ofSubtype_apply_of_mem _ (by exact ha)
+  · rw [iff_true_left _]
+    · rw [ofSubtype_apply_of_not_mem, ← not_mem_support]
+      · exact Finset.not_mem_mono (support_zpow_le c k) ha
+      exact ha
+    exact fun b ↦ (ha b).elim
+
+theorem zpow_eq_ofSubtype_subtypePerm_iff
+    {g c : Equiv.Perm α} {s : Finset α}
+    (hg : ∀ x, x ∈ s ↔ g x ∈ s) (hc : c.support ⊆ s) (n : ℤ) :
+    c ^ n = ofSubtype (g.subtypePerm hg) ↔
+      c.subtypePerm (isInvariant_of_support_le hc) ^ n = g.subtypePerm hg := by
+  constructor
+  · intro h; ext ⟨x, hx⟩; let h' := Perm.congr_fun h x
+    simp only [h', subtypePerm_apply_zpow_of_mem, Subtype.coe_mk, subtypePerm_apply]
```

Comment:
```suggestion
  · intro h; ext ⟨x, hx⟩
    simp only [Perm.congr_fun h x, subtypePerm_apply_zpow_of_mem, Subtype.coe_mk, subtypePerm_apply]
```
or at least `have` rather than `let`